File: src/main/java/space/serphantom/project/UserService.java
package space.serphantom.project;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class UserService {
    private Long nextId = 1L;
    private List<User> users = Collections.emptyList();

    public void save(final User user) {
        if (user.id() == 0) {
            User newUser = user.newBuilder().setId(nextId++).build();

            users = new ArrayList<>(users);
            users.add(newUser);
        } else {
            for (int i = 0; i < users.size(); i++) {
                if (users.get(i).id() == user.id()) {
                    users.set(i, user);

                    break;
                }
            }
        }
    }

    public void removeById(final long userId) {
        users = new ArrayList<>(users);
        users.removeIf(user -> user.id() == userId);
    }

    // Метод для получения пользователя по id
    public User getById(final long userId) {
        for (User user : users) {
            if (user.id() == userId) {
                return user;
            }
        }

        // The user by id numbers was not found
        throw new NotFoundException(
                "NotFoundException: " + "The User by id numbers: " + userId + " was not found."
        );
    }

    public List<User> getAll() {
        return Collections.unmodifiableList(users);
    }
}

File: src/main/kotlin/space/serphantom/project/UserService.kt
package space.serphantom.project

import java.util.*

class UserService {
    private var nextId: Long = 1
    private var users: MutableList<User> = Collections.emptyList()

    fun save(user: User) {
        if (user.id == 0L) {
            val newUser = user.newBuilder().setId(nextId++).build()

            users = ArrayList(users)
            users.add(newUser)
        } else {
            for (i in users.indices) {
                if (users[i].id == user.id) {
                    users[i] = user
                    break
                }
            }
        }
    }

    fun removeById(userId: Long) {
        users = ArrayList(users)
        users.removeIf { user: User -> user.id == userId }
    }

    fun getById(userId: Long): User {
        for (user in users) {
            if (user.id == userId) {
                return user
            }
        }

        // The user by id numbers was not found
        throw NotFoundException(
            "NotFoundException: The User by id numbers: $userId was not found."
        )
    }

    val all: List<User>
        get() = Collections.unmodifiableList(users)
}
